dist/diffhtml.js

```diff
@@ -2410,58 +2413,42 @@ function createPool(name, opts) {
     cache: cache,
 
     get: function get() {
-      var obj = null;
-      var freeLength = cache.free.length;
-      var minusOne = freeLength - 1;
-
-      if (freeLength) {
-        obj = cache.free[minusOne];
-        cache.free.length = minusOne;
-      } else {
-        obj = fill();
-      }
-
-      cache.allocated.push(obj);
-
-      return obj;
+      var value = cache.free.pop() || fill();
+      cache.allocated.add(value);
+      return value;
     },
     protect: function protect(value) {
-      var idx = cache.allocated.indexOf(value);
-
-      // Move the value out of allocated, since we need to protect this from
-      // being free'd accidentally.
-      if (cache.protected.indexOf(value) === -1) {
-        cache.protected.push(idx === -1 ? value : cache.allocated.splice(idx, 1)[0]);
-      }
+      cache.allocated.delete(value);
+      cache.protected.add(value);
 
-      // If we're protecting an element object, push the uuid into a lookup
-      // table.
       if (name === 'elementObject') {
         cache.uuid[value.uuid] = value;
       }
     },
     unprotect: function unprotect(value) {
-      var idx = cache.protected.indexOf(value);
-
-      if (idx !== -1) {
-        var obj = cache.protected.splice(idx, 1)[0];
-
-        if (obj) {
-          cache.free.push(obj);
-        }
+      if (cache.protected.has(value)) {
+        cache.protected.delete(value);
+        cache.free.push(value);
+      }
 
-        if (name === 'elementObject') {
-          delete cache.uuid[value.uuid];
-        }
+      if (name === 'elementObject') {
+        delete cache.uuid[value.uuid];
       }
     },
     freeAll: function freeAll() {
       var freeLength = cache.free.length;
       var minusOne = freeLength - 1;
-      var reAlloc = cache.allocated.slice(0, size - minusOne);
+
+      // All of this could go away if we could figure out `Array.from` within
+      // a PhantomJS web-worker.
+      var reAlloc = [];
+      cache.allocated.forEach(function (v) {
+        return reAlloc.push(v);
+      });
```

Comment:
A succinct way to rewrite this would be:

```js
var reAlloc = [...cache.allocated];
```

[This compiles to `[].concat()` though](http://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=var%20x%20%3D%20%5B1%2C2%2C3%5D%3B%0Avar%20y%20%3D%20%5B...x%5D%3B%0A%0Aconsole.log(y%20%3D%3D%3D%20x)%3B), which creates a dynamically sized array. I wonder how that compares speed-wise to the push system you've got going. In any event, I'd expect that fixing the size would be faster than both, but I'd be curious to see the perf diff.

```js
var reAlloc = new Array(cache.allocated.length);
cache.allocated.forEach((v, i) => reAlloc[i] = v);
```